test/common_device_type.py

```diff
@@ -0,0 +1,218 @@
+import inspect
+from functools import wraps
+import unittest
+import torch
+from common_utils import TestCase
+
+# Note: Generic Device-Type Testing
+#
+# [WRITING TESTS]
+#
+# Write your test class as usual except:
+#   (1) Only define test methods in the test class itself. Helper methods
+#       and non-methods must be inherited. This limitation is for Python2
+#       compatibility.
+#   (2) Each test method should have have the signature
+#           testX(self, device)
+#       The device argument will be a string like 'cpu' or 'cuda.'
+#   (3) Prefer using test decorators defined in this file to others.
+#       For example, using the @skipIfNoLapack decorator instead of the
+#       @skipCPUIfNoLapack will cause the test to not run on CUDA if
+#       LAPACK is not available, which is wrong. If you need to use a decorator
+#       you may want to ask about porting it to this framework.
+#
+#   See the TestTorchDeviceType class in test_torch.py for an example.
+#
+# [RUNNING TESTS]
+#
+# After defining your test class call instantiate_device_type_tests on it
+# and pass in globals() for the second argument. This will instantiate
+# discoverable device-specific test classes from your generic class. It will
+# also hide the tests in your generic class so they're not run directly.
+#
+# For each generic testX, a new test textX_<device_type>  will be created.
+# These tests will be put in classes named GenericTestClassName<DEVICE_TYPE>.
+# For example, test_diagonal in TestTorchDeviceType becomes test_diagonal_cpu
+# in TestTorchDeviceTypeCPU and test_diagonal_cuda in TestTorchDeviceTypeCUDA.
+#
+# In short, if you write a test signature like
+#   def textX(self, device)
+# You are effectively writing
+#   def testX_cpu(self, device='cpu')
+#   def textX_cuda(self, device='cuda')
+#   def testX_xla(self, device='xla')
+#   ...
+#
+# These tests can be run directly like normal tests:
+# "python test_torch.py TestTorchDeviceTypeCPU.test_diagonal_cpu"
```

Comment:
are we blocked on just calling these "TestTorchCPU" and similar?

Reply:
We can call them whatever you like but there already is a distinct TestTorch class. Maybe there won't be after we refactor?

Reply:
good question, but there isn't a TestTorchCPU class in any case :).